### PR TITLE
[SYCL] Enable basic support for 2020 specialization constants in sycl-post-link

### PR DESCRIPTION
**Pull request by vmaksimo**
SYCL 2020 specialization constants implementation will use new intrinsic in
DPC++ headers to provide required markup for the compiler.

By doing this change, SpecConstantsPass starts to support new interface for
specialization constants. It doesn't implement handling of new arguments, which
would be done in further PRs (for now new arguments are just ignored).

Spec is under review here, but it doesn't affect the patch promotion: https://github.com/intel/llvm/pull/3331

### CODE REVIEW

**Comment by romanovvlad** [comment]
@bader  @kbobrovs  Could you please review?

File: sycl/include/CL/sycl/detail/sycl_fe_intrins.hpp [hunk 1]
```diff
@@ -21,4 +21,14 @@ SYCL_EXTERNAL T __sycl_getSpecConstantValue(const char *ID);
 template <typename T>
 SYCL_EXTERNAL T __sycl_getCompositeSpecConstantValue(const char *ID);
 
+template <typename T>
+SYCL_EXTERNAL T __sycl_getScalar2020SpecConstantValue(const char *SymbolicID,
```
Comment (kbobrovs) [inline sycl/include/CL/sycl/detail/sycl_fe_intrins.hpp:25]: please add description and doc comments for each new intrinsic

File: llvm/test/tools/sycl-post-link/composite-spec-constant.ll [hunk 2]
```diff
@@ -23,7 +23,17 @@
 ; CHECK: %[[#POD:]] = call %struct._ZTS3POD.POD @"_Z29__spirv_SpecConstantCompositeAstruct._ZTS1A.Aclass._ZTSN2cl4sycl3vecIiLi2EEE.cl::sycl::vec"([2 x %struct._ZTS1A.A] %[[#NA]], %"class._ZTSN2cl4sycl3vecIiLi2EEE.cl::sycl::vec" %[[#B]]), !SYCL_SPEC_CONST_SYM_ID ![[#MD:]]
 ; CHECK: store %struct._ZTS3POD.POD %[[#POD]]
 ;
+; Test checks related to 2020 API for composite specialization constants.
```
Comment (kbobrovs) [inline llvm/test/tools/sycl-post-link/composite-spec-constant.ll:26]: this does not seem to have any 2020 specifics, am I missing something?

File: llvm/test/tools/sycl-post-link/composite-spec-constant.ll [hunk 3]
```diff
@@ -57,6 +69,30 @@ entry:
...
+  %4 = load i32, i32 addrspace(4)* %a.i, align 4
+  %conv.i = sitofp i32 %4 to float
+  %b.i = getelementptr inbounds %struct._ZTS1A.A, %struct._ZTS1A.A addrspace(4)* %c.ascast.i, i64 0, i32 1
+  %5 = load float, float addrspace(4)* %b.i, align 4
+  %add.i = fadd float %5, %conv.i
+  %ptridx.ascast.i.i = addrspacecast float addrspace(1)* %add.ptr.i to float addrspace(4)*
+  store float %add.i, float addrspace(4)* %ptridx.ascast.i.i, align 4, !tbaa !11
+  call void @llvm.lifetime.end.p0i8(i64 8, i8* nonnull %3) #3
+  ret void
+}
```
Comment (kbobrovs) [inline llvm/test/tools/sycl-post-link/composite-spec-constant.ll:94]: there no single CHECK in this function - what is the purpose of adding it?

Comment (vmaksimo) [reply llvm/test/tools/sycl-post-link/composite-spec-constant.ll:94]: This function was added to test new `__sycl_get*2020SpecializationConstant` intrinsic (which are introduced in spec constants 2020 design).
The corresponding checks for this piece of IR are located at the beginning of the file (as it was done for this test initially).

File: llvm/test/tools/sycl-post-link/composite-spec-constant.ll (hunk 2 above)
Comment (vmaksimo) [reply llvm/test/tools/sycl-post-link/composite-spec-constant.ll:26]: These checks are just related to function with new intrinsic API calls, but you're right that IR doesn't look different.
I'll remove this mark as a bit confusing one.

File: llvm/test/tools/sycl-post-link/composite-spec-constant.ll (hunk 3 above)
Comment (kbobrovs) [reply llvm/test/tools/sycl-post-link/composite-spec-constant.ll:94]: I suggest to move the checks next to the source IR portion which are supposed to produce the checked result. Otherwise it is really hard to follow the logic of the test.

Comment (vmaksimo) [reply llvm/test/tools/sycl-post-link/composite-spec-constant.ll:87]: Agree, done

File: llvm/test/tools/sycl-post-link/composite-spec-constant.ll [hunk 4]
```diff
@@ -69,6 +57,11 @@ entry:
   ret void
 }
 
+; CHECK: %[[#N0:]] = call i32 @_Z20__spirv_SpecConstantii(i32 [[#ID + 6]], i32
```
Comment (kbobrovs) [inline llvm/test/tools/sycl-post-link/composite-spec-constant.ll:60]: My idea was to actually move the checks right below the `__sycl_get*` intrinsics which lead to the checked IR.

File: sycl/include/CL/sycl/detail/sycl_fe_intrins.hpp [hunk 5]
```diff
@@ -21,6 +21,16 @@ SYCL_EXTERNAL T __sycl_getSpecConstantValue(const char *ID);
 template <typename T>
 SYCL_EXTERNAL T __sycl_getCompositeSpecConstantValue(const char *ID);
 
+// The intrinsics below are used to enable support of specialization constants
+// 2020. This feature required post-link to handle more parameters.
```
Comment (kbobrovs) [inline sycl/include/CL/sycl/detail/sycl_fe_intrins.hpp:25]: Thanks! One more nit

```suggestion
// The intrinsics below are used to implement support SYCL2020 specialization constants.
// SYCL2020 version requires more parameters compared to the initial version.
```

File: llvm/test/tools/sycl-post-link/spec_const_O2.ll [hunk 6]
```diff
@@ -180,3 +180,4 @@ attributes #1 = { "correctly-rounded-divide-sqrt-fp-math"="false" "disable-tail-
 ; CHECK-RT: ![[ID8]] = !{!"_ZTS13MyUInt64Const", i32 8}
 ; CHECK-RT: ![[ID9]] = !{!"_ZTS12MyFloatConst", i32 9}
 ; CHECK-RT: ![[ID10]] = !{!"_ZTS13MyDoubleConst", i32 10}
+; CHECK-RT: ![[ID11]] = !{!"_ZTS14MyDoubleConst2", i32 11}
```
Comment (kbobrovs) [inline llvm/test/tools/sycl-post-link/spec_const_O2.ll:183]: add eof?

**Review by AlexeySachkov — APPROVED** [review]
LGTM modulo comments from Konst (the one about doc comment for each new intrinsic is still haven't addressed)

**Comment by vmaksimo** [comment]
> LGTM modulo comments from Konst (the one about doc comment for each new intrinsic is still haven't addressed)

I don't think that we need to duplicate the description for composite intrinsic as the comment would be a copy-paste. It relates to both of the new intrinsics (the same way that a previous couple of intrinsics was documented above).

**Comment by bader** [comment]
@kbobrovs, please, take a look if your concerns are resolved.

**Comment by AlexeySachkov** [comment]
> > LGTM modulo comments from Konst (the one about doc comment for each new intrinsic is still haven't addressed)
> 
> I don't think that we need to duplicate the description for composite intrinsic as the comment would be a copy-paste. It relates to both of the new intrinsics (the same way that a previous couple of intrinsics was documented above).

Ok, works for me

**Review by kbobrovs — APPROVED** [review]

**Comment by kbobrovs** [comment]
> > > LGTM modulo comments from Konst (the one about doc comment for each new intrinsic is still haven't addressed)
> > 
> > 
> > I don't think that we need to duplicate the description for composite intrinsic as the comment would be a copy-paste. It relates to both of the new intrinsics (the same way that a previous couple of intrinsics was documented above).
> 
> Ok, works for me

For met too. Added comment seems enough.

**Comment by kbobrovs** [comment]
> @kbobrovs, please, take a look if your concerns are resolved.

@bader, yes, approved.
